### PR TITLE
Change to a Pin-based connection model.

### PR DESCRIPTION
**Pull request by Grissess**
This commit separates the logic of connection from the nodes used in
nodal analysis. The combination of the two was a byproduct of an older
API and causing misconceptions in the correct usage; errant usages are
much easier to obviate. As a result, connection to ground is now
special-cased.

### CODE REVIEW

File: core/src/main/kotlin/org/eln2/debug/MatrixPrint.kt [hunk 1]
```diff
@@ -73,54 +76,7 @@ fun matrixFormat(matrix: RealMatrix, headerfooter: Boolean = true): String {
  * @return matrix output
  */
 fun knownsFormat(knowns: RealVector, headerfooter: Boolean = true): String {
```
[the remark this answers is not in the document]
Comment (jrddunbr) [reply core/src/main/kotlin/org/eln2/debug/MatrixPrint.kt:78]: I know a RealVector != RealMatrix, but just overload it with null checks.

File: core/src/main/kotlin/org/eln2/sim/electrical/mna/Circuit.kt [hunk 2]
```diff
@@ -341,10 +339,8 @@ class Circuit {
         componentsChanged = true
         // This is the ONLY place where this should be set.
         comp.circuit = this
-        for (i in 0 until comp.nodeCount) {
-            val n = Node(this)
-            compNodeMap[n] = comp
-            comp.nodes.add(n)
+        for (i in 0 until comp.pinCount) {
```
Comment (jrddunbr) [inline core/src/main/kotlin/org/eln2/sim/electrical/mna/Circuit.kt:342]: Replace with `(0 .. until comp.pinCount).forEach{}` style call.

File: core/src/main/kotlin/org/eln2/sim/electrical/mna/Circuit.kt [hunk 3]
```diff
@@ -341,10 +339,8 @@ class Circuit {
...
+            comp.pins.add(Pin())
         }
         for (i in 0 until comp.vsCount) {
```
Comment (jrddunbr) [inline core/src/main/kotlin/org/eln2/sim/electrical/mna/Circuit.kt:345]: forEach here too

File: core/src/main/kotlin/org/eln2/sim/electrical/mna/Circuit.kt [hunk 4]
```diff
@@ -416,19 +413,87 @@ class Circuit {
...
+        var i = 0
+        for(p in pinSet) {
+            if(p.node == null) {
+                p.node = Node(this)
+            }
+            if(p.node != ground) {
+                p.node!!.index = i++
+                newNodes.add(WeakReference(p.node))
+            }
+        }
```
Comment (jrddunbr) [inline core/src/main/kotlin/org/eln2/sim/electrical/mna/Circuit.kt:437]: This hurts my bones.

File: core/src/main/kotlin/org/eln2/sim/electrical/mna/component/Capacitor.kt [hunk 5]
```diff
@@ -15,20 +15,15 @@ open class Capacitor : Port() {
     override var name: String = "c"
 
     /**
-     * The current, in Amperes, presently sourced by this Norton system.
+     * Capacitance, in Farads (Coulombs / Volt)
      */
-    internal var i: Double = 0.0
+    var capacitance: Double = 1e-5
         set(value) {
-            if (isInCircuit)
```
Comment (jrddunbr) [inline core/src/main/kotlin/org/eln2/sim/electrical/mna/component/Capacitor.kt:22]: I guess we don't need this?

File: core/src/test/kotlin/org/eln2/sim/electrical/mna/CircuitTest.kt [hunk 6]
```diff
@@ -19,11 +19,12 @@ internal class CircuitTest {
...
             c.add(vs, r1)
-            vs.neg.named("vneg")
-            vs.pos.named("vpos")
+            // TODO: figure out how to name nodes early again
+            // vs.neg.named("vneg")
+            // vs.pos.named("vpos")
             vs.connect(1, r1, 1)
             vs.connect(0, r1, 0)
-            vs.connect(1, c.ground)
+            vs.ground(1)
```
Comment (jrddunbr) [inline core/src/test/kotlin/org/eln2/sim/electrical/mna/CircuitTest.kt:27]: +1 I like this syntax better.

File: core/src/test/kotlin/org/eln2/sim/electrical/mna/CircuitTest.kt [hunk 7]
```diff
@@ -106,14 +108,14 @@ internal class CircuitTest {
         r2.connect(0, ts.r1, 0)
         r2.connect(1, ts.r1, 1)
 
-        ts.c.step(0.5)
+        assert(ts.c.step(0.5))
```
Comment (jrddunbr) [inline core/src/test/kotlin/org/eln2/sim/electrical/mna/CircuitTest.kt:111]: what does this assert do? It's not extremely clear

File: core/src/test/kotlin/org/eln2/sim/electrical/mna/CircuitTest.kt [hunk 8]
```diff
@@ -187,19 +189,20 @@ internal class CircuitTest {
...
 
         for (comp in c.components) {
-            println("main_1: comp $comp name ${comp.name} nodes ${comp.nodes} vs ${comp.vsources}")
-            for (node in comp.nodes) {
-                println("\t node $node index ${node.index} ground ${node.isGround} potential ${node.potential}")
+            println("main_1: comp $comp name ${comp.name} pins ${comp.pins} vs ${comp.vsources}")
+            for (pin in comp.pins) {
+                val node = pin.node
+                println("\t node $node index ${node?.index} ground ${node?.isGround} potential ${node?.potential}")
             }
```
Comment (jrddunbr) [inline core/src/test/kotlin/org/eln2/sim/electrical/mna/CircuitTest.kt:206]: forEach()

**Review by jrddunbr — CHANGES_REQUESTED** [review]
Just some minor nitpicks. There is one function where my bones hurt a lot and I can't think of a way to make it less C++ like.

Also, if you're going to make it `mnaPrint()`, perhaps do `mnaPrintln()` as well?

**Comment by jrddunbr** [comment]
Holy shit the SNR on this new linter is shit.

File: core/src/main/kotlin/org/eln2/debug/MatrixPrint.kt [hunk 9]
```diff
@@ -16,53 +15,57 @@ const val FORMAT_SIZE = 8
...
+    val singleRow = rows <= 1
+    val sb = StringBuilder()
+
+    if (headerfooter) sb.append("== Begin MNA Matrix ==\n")
+    
+    val columnWidths = (0 until columns).map {
+        matrix.getColumn(it).map { entry -> entry.toString().length }.max() ?: 0
+    }.toList()
+
+    for(row in 0 until rows) {
```
[the remark this answers is not in the document]
Comment (Grissess) [reply core/src/main/kotlin/org/eln2/debug/MatrixPrint.kt:30]: I'd been advised against that by [this Codacy lint](https://arturbosch.github.io/detekt/performance.html#foreachonrange) for performance reasons. Supposedly, that will be fixed eventually, but until then...

File: core/src/main/kotlin/org/eln2/debug/MatrixPrint.kt (hunk 1 above)
[the remark this answers is not in the document]
Comment (Grissess) [reply core/src/main/kotlin/org/eln2/debug/MatrixPrint.kt:78]: Yeah, that can be collapsed :)

File: core/src/main/kotlin/org/eln2/sim/electrical/mna/Circuit.kt (hunk 2 above)
Comment (Grissess) [reply core/src/main/kotlin/org/eln2/sim/electrical/mna/Circuit.kt:342]: See the performance lint above.

File: core/src/test/kotlin/org/eln2/sim/electrical/mna/CircuitTest.kt (hunk 7 above)
Comment (Grissess) [reply core/src/test/kotlin/org/eln2/sim/electrical/mna/CircuitTest.kt:111]: `step` returns whether or not it succeeds; the most typical reason for it failing is encountering a singular matrix. If we're assuming that's not going to happen, asserting this is a good way to catch errors earlier (before the numerical asserts trigger).

File: core/src/test/kotlin/org/eln2/sim/electrical/mna/CircuitTest.kt (hunk 8 above)
Comment (Grissess) [reply core/src/test/kotlin/org/eln2/sim/electrical/mna/CircuitTest.kt:206]: eh... yeah, that one can be done...

File: core/src/main/kotlin/org/eln2/debug/MatrixPrint.kt (hunk 9 above)
[the remark this answers is not in the document]
Comment (jrddunbr) [reply core/src/main/kotlin/org/eln2/debug/MatrixPrint.kt:30]: lol, k

File: core/src/test/kotlin/org/eln2/sim/electrical/mna/CircuitTest.kt (hunk 7 above)
Comment (jrddunbr) [reply core/src/test/kotlin/org/eln2/sim/electrical/mna/CircuitTest.kt:111]: Ah, I see. Nice.

File: core/src/main/kotlin/org/eln2/sim/electrical/mna/component/Capacitor.kt (hunk 5 above)
Comment (jrddunbr) [reply core/src/main/kotlin/org/eln2/sim/electrical/mna/component/Capacitor.kt:22]: by "this" I mean `var i`

Comment (Grissess) [reply core/src/main/kotlin/org/eln2/sim/electrical/mna/component/Capacitor.kt:22]: It was subsumed, to my understanding, by the current field, as intended. I'm trying to get away from variables and fields names for math symbols.

File: core/src/test/kotlin/org/eln2/sim/electrical/mna/CircuitTest.kt (hunk 6 above)
Comment (Grissess) [reply core/src/test/kotlin/org/eln2/sim/electrical/mna/CircuitTest.kt:27]: It really was the only special case; that should have happened forever ago :P

**Comment by Grissess** [comment]
I'm not sure why IDEA kept insisting on globbing the `java.util` import, but it really _was_ just `WeakHashMap`. I wonder if that's a setting...

@jrddunbr ready for next check :)

File: core/src/main/kotlin/org/eln2/sim/electrical/mna/Circuit.kt (hunk 4 above)
Comment (jrddunbr) [reply core/src/main/kotlin/org/eln2/sim/electrical/mna/Circuit.kt:437]: My bones!!

File: core/src/main/kotlin/org/eln2/sim/electrical/mna/component/Capacitor.kt (hunk 5 above)
Comment (jrddunbr) [reply core/src/main/kotlin/org/eln2/sim/electrical/mna/component/Capacitor.kt:22]: Then how are we keeping track of the current source's current? `current` had the *overall* current over the item, and `i` had the current over the *current source* itself. Has that changed?

**Pull request merged**